### ISSUE TITLE
Hang when exiting subshell

### ISSUE THREAD

**Issue by dnicolson** [fish-shell/fish-shell#7060]
If the following code is run as `go run subshell.go`, the terminal session in iTerm2 can become unresponsive and cause the `fish` process to run at 99% CPU.
```go
package main

import "os"
import "os/exec"

func main() {
	cmd := exec.Command(os.Getenv("SHELL"), "--login")
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin
	cmd.Run()
}
```

1. `go run subshell.go`
2. Hit Control-C
3. Hit enter
4. Hit enter

This results in the following output before the terminal session becomes unresponsive:
```
$ go run subshell.go
$ signal: interrupt
$ ⏎
$
$
$
```

The output of the Terminal app is similar, except the terminal session closes and fish stops running.

Performing the above steps with fish 2.7.1 does not have the same issue.

fish 3.1.2
Darwin Kernel Version 19.5.0: root:xnu-6153.121.1~7/RELEASE_X86_64 x86_64
xterm-256color
iTerm 3.3.11beta1

**Comment by mqudsi** [comment]
That just launches a bash environment for me and nothing breaks (tested under conhost and kitty). 

What is your `$SHELL`?

**Comment by dnicolson** [comment]
$SHELL is `/usr/local/bin/fish`.

**Comment by mqudsi** [comment]
Not reproducing for me under Linux.

![image](https://user-images.githubusercontent.com/606923/83317584-6a412300-a1f3-11ea-9c49-077725138864.png)

**Comment by mqudsi** [comment]
Wait, it seems to be an invalid pgroup condition. In that same environment, after exiting the subshell started by go, if I just try to rerun the command again I get the following:

![image](https://user-images.githubusercontent.com/606923/83317625-a5435680-a1f3-11ea-8845-944f89c089de.png)


This is yet another terminal pgroup bug, and probably bisects to the same changes we've seen before.

**Comment by mqudsi** [comment]
I was able to reproduce this and then suddenly I wasn't any more. I believe it's a race condition with the terminal owning pgrp (which is the general issue with our pgrp bugs), as I got the following error one of the times that it reproduced for me:

![image](https://user-images.githubusercontent.com/606923/83318703-cf4d4680-a1fc-11ea-8ee7-fc535c660028.png)

And now that I have all my ducks in a row to be able to debug this, it won't reproduce for me any more. Go figure.

I still had my default `$SHELL` set to `/usr/bin/bash` on one of the times that I was able to reproduce it and the bash session that it launched had incorrect terminal flags (it seemed that ECHO was off) *but* could read and write from the terminal OK (commands were still executed as blindly typed).

**Comment by mqudsi** [comment]
Ok, I think this one is caused by the fact that fish doesn't forcibly create a new process group with control of the terminal. When you run the go code in your shell, the shell creates a new process group for the `go` process. `go` is the only command in this process group and is therefore the process group leader.

When go launches fish, its a naive execution that doesn't do any form of job control. fish is launched into go's process group and inherits all its handles. The `go` process is still the process group leader, and the process group has full control of the terminal. 

When the `go` process exits (when you hit ^C), the process group is now orphaned (it has no members whose parents are members of another process group in the same session session). Whatever shell you ran that command in (bash, fish, etc) was waiting for `go` to exit so that it could retake control of the terminal. And retake control of the terminal it does, since it was the parent of that process group leader.

Two different issues are coming into play here:
1) fish doesn't create a new process group for itself on startup (except under some exceptional circumstances, e.g. firejail) so the shell that spawned `go` is free to regain control of the terminal
2) when `go` exits and is reaped by *its* parent and the subshell `fish` becomes a member of an orphaned process group that no longer has control of the terminal, its call to `read`/`readch` from STDIN triggers a SIGTTIN (attempting to read from a tty it does not control). The fish sighandler catches this and basically does nothing fruitful about it, ultimately letting the `read` call continue after failing (with `EIO`, I believe) immediately afterwards.

The first point is somewhat deliberate, although it might be worth revisiting given the code cleanup that we've done since that time. The basic gist is that by not creating our own process group, we assume that we are being launched by a shell or something else smart enough to do its own job control. But under that assumption, the right thing to do with a SIGTTIN would be to suspend (SIGSTOP) ourselves until we've been given control of the terminal again (not sure how that would happen apart from being directly `fg`'d by the user, but that is the assumption). If we didn't override the SIGTTIN or we handled it by cleaning up and then pausing, it would be OK (for some definition of that word): the fish subshell, no longer in control of the terminal and not in the foreground, would stop trying to read from STDIN or write to STDOUT and CPU usage would remain at 0.

But what ends up happening is that we (similarly no longer in control of the terminal and similarly no longer in the foreground) ignore the SIGTTIN, keep trying to read from STDIN, keep getting SIGTTIN, rinse and repeat - i.e. the worst of both worlds.

**Comment by ridiculousfish** [comment]
See this [blast from the past](https://stackoverflow.com/questions/13718394/what-should-interactive-shells-do-in-orphaned-process-groups).

fish has special logic to try to notice when it is orphaned. Perhaps it needs to be updated.

**Comment by mqudsi** [comment]
Ah yes, that SO post. I remember coming across it many moons ago!

One option, even if not the most elegant, is to just SIGHUP; logically the situation isn’t too different from a terminal window closing on us, as are the chances of regaining control of the tty. (We wouldn’t do this automatically on SIGTTIN, we’d have to first ascertain we’ve been actually orphaned, otherwise we could have been backgrounded instead - in which case we should probably SIGSTOP until we have control of the terminal again).

**Comment by ridiculousfish** [comment]
This is pretty amusing:

1. fish launches go launches fish. Call these "fish1" and "fish2".  go and fish2 are in the same process group.
2. SIGINT is delivered. go exits. fish2 is interactive, it swallows SIGINT. Now fish2 is orphaned.
3. fish1 sees that go has exited and reclaims the terminal.
4. fish2 soon after decides to exit gracefully - probably it thinks stdin is closed. One of the last things it does is "helpfully" restore the foreground process group to what it initially saw, with is go's orphaned group.
5. This in turn steals the tty from fish1 without fish1 noticing. So we get the SIGTTIN.

The right fixes here are:

1. Probably fish should not restore the foreground process group if it is orphaned. But it's hard to know that for sure - probably `kill(0, getpgrp())` and check for ESRCH.

2. Definitely fish should be better about handling SIGTTIN, but it's not obvious how. The usual approach when you get SIGTTIN is to wait until you have ownership of the terminal, but here ownership won't ever arrive.

**Comment by mqudsi** [comment]
See my comment above about alternative options. We were posting simultaneously. 

I think an easy win is to use tcgetpgrp before using tcsetpgrp. If we’re no longer in control of the tty (pgrp doesn’t match ours or we get ENOTTY) then we shouldn’t issue a tcsetpgrp. (I also don’t see how tcsetpgrp can succeed if “fish2” issues it and doesn’t have control of the tty and is merely a process in the same session?)

**Comment by ridiculousfish** [comment]
> I also don’t see how tcsetpgrp can succeed if “fish2” issues it and doesn’t have control of the tty and is merely a process in the same session?

Heh, it's a bit of Unix sillyness. `tcsetpgrp` is allowed by a non-foreground proc if `SIGTTOU` is blocked, which fish does. So then that's the right idea here - unblock SIGTTOU and then tcsetpgrp will fail. Clever.

**Comment by ridiculousfish** [comment]
Fixed as 3ae91f197da82b60780d3aacca21bcb3f11aad87. Awesome debugging @mqudsi! Process groups sure are "fun".

**Comment by ridiculousfish** [comment]
Also thanks to @dnicolson for filing.

**Comment by mqudsi** [comment]
Wow, that was fast! 👍

I do wish man pages were more explicit:

> If  tcsetpgrp()  is  called  by  a  member of a background process group in its session, and the calling process is not blocking or ignoring SIGTTOU, a SIGTTOU signal is sent to all members of this  background process group.

What I'm *not* seeing is explicit mention that the call will actually fail, in addition to the SIGTTOU being sent. 

More worryingly, FreeBSD's man page makes no mention of the SIGTTOU exception:

> If the process has a controlling terminal, the tcsetpgrp() function sets the foreground process group ID associated with the terminal device to pgrp_id. The terminal device associated with fd must be the controlling terminal of the calling process and the controlling terminal must be currently associated with the session of the calling process. The value of pgrp_id must be the same as the process group ID of a process in the same session as the calling process.

>     The tcsetpgrp() function will fail if:                                                                                                                                                                                              
> *[EBADF]            The fd argument is not a valid file descriptor.                                                                                                                                                                 
> *[EINVAL]           An invalid value of pgrp_id was specified.                                                                                                                                                                      
> *[ENOTTY]           The calling process does not have a controlling terminal, or the file represented by fd        is not the controlling terminal, or the controlling terminal is no longer associated with the session of the      calling process.                                                                                                                                                                                                                   
> *[EPERM]            The pgrp_id argument does not match the process group ID of a process in the same session      as the calling process.

**Comment by mqudsi** [comment]
(Ugh. I hate to be a Debbie Downer, but it doesn't seem to be fixed on WSL/Linux or FreeBSD.)

**Comment by mqudsi** [comment]
This seems to work

```diff
     if (initial_fg_process_group != -1) {
         // This is called during shutdown and from a signal handler. We don't bother to complain on
         // failure because doing so is unlikely to be noticed.
-        // However we want this to fail if we are not the tty owner (#7060), so clear our SIGTTOU
-        // handler to allow it to fail properly. Note that we are about to exit.
-        (void)signal(SIGTTOU, SIG_DFL);
-        (void)tcsetpgrp(STDIN_FILENO, initial_fg_process_group);
+        // However we want this to fail if we are not the tty owner (#7060), so check with tcgetpgrp
+        // before issuing tcsetpgrp.
+        if (tcgetpgrp(STDIN_FILENO) == getpgrp()) {
+            (void)tcsetpgrp(STDIN_FILENO, initial_fg_process_group);
+        }
     }
 }
```

**Comment by ridiculousfish** [comment]
Yeah WSL is probably different.  I forget where I learned about the creepy-action-at-a-distance of SIGTTOU <-> tcsetpgrp but it's basically there just so shells don't break.

**Comment by dnicolson** [comment]
@ridiculousfish that change fixed the original example.

I am still getting hangs though, it looks like the problem relates to binaries:

1. `go build subshell.go`
2. `./subshell`
3. Hit up
4. Hit Control-C
5. Hit enter

**Comment by ridiculousfish** [comment]
@dnicolson I cannot reproduce hangs with your latest steps. Is it possible that "$SHELL" is referring to a fish without the fix? Perhaps you have not yet installed from master?

**Comment by dnicolson** [comment]
I think the original commands I used to install the latest fish from master were:
```
$ brew unlink fish
$ brew install fish --HEAD
```

```
$ echo $SHELL
/usr/local/bin/fish
$ /usr/local/bin/fish -v
fish, version 3.1.2-771-g3ae91f197
```

Here is an asciinema recording, although the problem doesn't exist when recording:
https://asciinema.org/a/xk7Xg2VcDKITYOiDJxcE2WRIs

Different terminals give different results:
- iTerm2 results in hung fish process
- kitty just closes the window
- Terminal shows some history/session output then closes the session

All terminals perform correctly if step 3 is omitted.

**Comment by mqudsi** [comment]
I can't reproduce it on FreeBSD but can on Fedora (I'm guessing @ridiculousfish tried it on a mac). 

It demonstrates the same behavior WSL was displaying, bouncing between `SIGTTIN` and `read`.

**Comment by mqudsi** [comment]
This branch checks for SIGTTIN (`sigint_checker_t` adapted for use with other signals) and fixes the problem with the inner fish instance getting stuck in SIGTTIN, but I think it breaks the parent fish instance (which gets the SIGTTIN but previously used to ignore it, I think?): https://github.com/mqudsi/fish-shell/tree/sigttin-checks

**Comment by ridiculousfish** [comment]
@mqudsi your proposed changes look great to me!

**Comment by mqudsi** [comment]
@ridiculousfish thanks, but unfortunately it breaks the host fish session somehow; as mentioned it too gets a SIGTTIN and this patch stops it from being able to pretend it didn't happen. It was 2 am though, so I called it a night.

The parent fish session is supposed to be stuck waiting for the child to exit and when ctrl-c is pressed, it takes that as a signal to resume execution. But it's actually not in control of the terminal because the spawned fish instance is still running in the pgrp that was given control of the tty. In particular, `is_foreground()` in proc.cpp returns true (because the job was created in the foreground) but the checks on lines 1030-1035 only enumerate processes known to exist (e.g. created by fish):

https://github.com/fish-shell/fish-shell/blob/4dff15b74ef46455b1eeaf5e55f2e2ae25232628/src/proc.cpp#L1029-L1035

We assume this is OK because `term_steal()` is called which regains control of the tty (which causes SIGTTIN to ultimately fire in the child, because it's been forcibly stripped of tty control), but the sequence of keys pressed by the user creates a race condition where fish1 detects `go` exited and calls `term_steal()` but fish2 starts a new loop to redraw the prompt when ctrl-c is *not* a no-op, taking control of the tty back! They race for control of the terminal, and even though fish2 gives control back to fish1 before exiting, it's too late because fish1 has *also* run into SIGTTIN (because ownership of the ctty changed in between the call to `term_steal()` and the call to `read()`) and forces itself to close (returning the controlling tty to init/sshd/whatever). 

Before the SIGTTIN patch in branch, both fish sessions would consume 100% cpu while bouncing between read/eint and SIGTTIN but the parent fish would break out of that tight loop when the child fish instance lost the tussle for control. With the patch, they both surrender just one of them too soon.

**Comment by ridiculousfish** [comment]
Nice detective work. We should understand whether/how bash handles this case.

**Comment by dnicolson** [comment]
On macOS Big Sur (20A5343i) the behaviour is a little different, the session terminates now instead of hanging with v3.1.2.

Using the latest version (3.1.2-1186-g86b02278b), on both Catalina and Big Sur random characters are output at the end:

```
$ go run subshell.go
$ signal: interrupt
$
[?2004l
```

Should this be a new issue?

**Comment by faho** [comment]
>random characters are output at the end:

Those aren't random characters. The disable sequence for bracketed paste is `\e[?2004l`, so most likely something is swallowing the escape (`\e`), causing the rest to be displayed.

**Comment by dnicolson** [comment]
I'm not sure if it's entirely random, but I was able to produce other characters such as:
`?3200m0(lB⏎`
`??22000044lh`
`04l`

**Comment by faho** [comment]
>I'm not sure if it's entirely random, but I was able to produce other characters such as:

Those are scrambled escape sequences.

Colors look like `\e[30m` (black in this case), "reset" is `\e(B\e[m`.

This seems like something is still reading here.

**Comment by ridiculousfish** [comment]
So we still have a race - this time it's between the child fish responding to SIGINT, and the parent fish attempting to reclaim the tty. What ends up happening is that child fish mucks with the tty modes and then exits, and parent fish is left with a bad tty.

I think the simplest workaround here is to just arrange for fish to be in its own pgroup - that's what happens with bash.

**Comment by ridiculousfish** [comment]
That's what I did. It's no good trying to run interactively in someone else's pgroup.

**Comment by faho** [comment]
This could also fix #5699.
